Dapper/DynamicParameters.cs

```diff
@@ -128,6 +150,32 @@ public void Add(string name, object value = null, DbType? dbType = null, Paramet
             {
                 Name = name,
                 Value = value,
+                Type = value?.GetType(),
+                ParameterDirection = direction ?? ParameterDirection.Input,
+                DbType = dbType,
+                Size = size,
+                Precision = precision,
+                Scale = scale
+            };
+        }
+
+        /// <summary>
+        /// Add a parameter to this dynamic parameter list.
+        /// </summary>
+        /// <param name="name">The name of the parameter.</param>
+        /// <param name="value">The value of the parameter.</param>
+        /// <param name="dbType">The type of the parameter.</param>
+        /// <param name="direction">The in or out direction of the parameter.</param>
+        /// <param name="size">The size of the parameter.</param>
+        /// <param name="precision">The precision of the parameter.</param>
+        /// <param name="scale">The scale of the parameter.</param>
+        public void Add<T>(string name, T value = default(T), DbType? dbType = null, ParameterDirection? direction = null, int? size = null, byte? precision = null, byte? scale = null)
+        {
+            parameters[Clean(name)] = new ParamInfo
+            {
+                Name = name,
+                Value = value,
+                Type = value?.GetType() ?? typeof(T),

```

Comment:
It *feels* like this should be `typeof(T)` always, since we're explicitly supplying the type information. In the case of a base type and a child type, `T` can be the base type and `value` can be of the child type in a valid generic call...so these can differ.

Thoughts?